### PR TITLE
Adding reporting on pod failures

### PR DESCRIPTION
**Pull request by jpodivin**

### CODE REVIEW

**Comment by jpodivin** [comment]
recheck

**Comment by jpodivin** [comment]
recheck
`Exception '403\nReason: Forbidden\nHTTP response headers: HTTPHeaderDict({'Audit-Id': '6b7b476c-e41a-4aca-ad33-332ba0c28df8', 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains; preload', 'X-Content-Type-Options': 'nosniff', 'X-Kubernetes-Pf-Flowschema-Uid': '', 'X-Kubernetes-Pf-Prioritylevel-Uid': '', 'Date': 'Mon, 22 Jan 2024 16:15:04 GMT', 'Content-Length': '259'})\nHTTP response body: b'{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"pods is forbidden: User \\\\\"kubeadmin\\\\\" cannot list resource \\\\\"pods\\\\\" in API group \\\\\"\\\\\" in the namespace \\\\\"cert-manager\\\\\"\",\"reason\":\"Forbidden\",\"details\":{\"kind\":\"pods\"},\"code\":403}\\n'\nOriginal traceback: \n  File \"/home/zuul/.local/lib/python3.9/site-packages/kubernetes/dynamic/client.py\", line 55, in inner\n    resp = func(self, *args, **kwargs)\n\n  File \"/home/zuul/.local/lib/python3.9/site-packages/kubernetes/dynamic/client.py\", line 270, in request\n    api_response = self.client.call_api(\n\n  File \"/home/zuul/.local/lib/python3.9/site-packages/kubernetes/client/api_client.py\", line 348, in call_api\n    return self.__call_api(resource_path, method,\n\n  File \"/home/zuul/.local/lib/python3.9/site-packages/kubernetes/client/api_client.py\", line 180, in __call_api\n    response_data = self.request(\n\n  File \"/home/zuul/.local/lib/python3.9/site-packages/kubernetes/client/api_client.py\", line 373, in request\n    return self.rest_client.GET(url,\n\n  File \"/home/zuul/.local/lib/python3.9/site-packages/kubernetes/client/rest.py\", line 244, in GET\n    return self.request(\"GET\", url,\n\n  File \"/home/zuul/.local/lib/python3.9/site-packages/kubernetes/client/rest.py\", line 238, in request\n    raise ApiException(http_resp=r)\n' raised while trying to get resource using (name=None, namespace=cert-manager, label_selectors=['app = cainjector'], field_selectors=[])"`

File: controllers/openstackdataplanenodeset_controller.go [hunk 1]
```diff
@@ -388,6 +389,21 @@ func (r *OpenStackDataPlaneNodeSetReconciler) SetupWithManager(mgr ctrl.Manager)
 			}
 			result = append(result, reconcile.Request{NamespacedName: name})
 		}
+		podsInterface := r.Kclient.CoreV1().Pods(namespace)
+		// List the pods in the given namespace
+		podsList, err := podsInterface.List(context.TODO(), v1.ListOptions{})
```
Comment (fao89) [inline controllers/openstackdataplanenodeset_controller.go:394]: you can filter by deployment: https://github.com/openstack-k8s-operators/dataplane-operator/blob/main/pkg/util/ansible_execution.go#L68

and you probably need to add it:
```
//+kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch
```
https://github.com/openstack-k8s-operators/dataplane-operator/blob/main/controllers/openstackdataplanenodeset_controller.go#L120

Comment (fao89) [reply controllers/openstackdataplanenodeset_controller.go:394]: you can also use the `field-selector` https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/ to filter `status.phase=Failed`

Comment (jpodivin) [reply controllers/openstackdataplanenodeset_controller.go]: Thanks, that does make sense.

File: controllers/openstackdataplanenodeset_controller.go [hunk 2]
```diff
@@ -388,6 +390,23 @@ func (r *OpenStackDataPlaneNodeSetReconciler) SetupWithManager(mgr ctrl.Manager)
 			}
 			result = append(result, reconcile.Request{NamespacedName: name})
 		}
+		podsInterface := r.Kclient.CoreV1().Pods(namespace)
+		// List service pods in the given namespace
+		podsList, err := podsInterface.List(context.TODO(), v1.ListOptions{
+			LabelSelector: "app.kubernetes.io/name=openstackdataplaneservice",
+		})
```
Comment (fao89) [inline controllers/openstackdataplanenodeset_controller.go:397]:
```suggestion
		podsList, err := podsInterface.List(context.TODO(), v1.ListOptions{
			LabelSelector: fmt.Sprintf("osdpd=%s",deployment.Name),
			FieldSelector: "status.phase=Failed",
		})
```

**Review by fao89 — APPROVED** [review]
Thank you!

**Comment by fao89** [comment]
Failed to push image: trying to reuse blob sha256:07a64a71e01156f8f99039bc246149925c6d1480d3957de78510bbec6ec68f7a at destination: pinging container registry quay.rdoproject.org


/test dataplane-operator-build-deploy-kuttl

**Comment by fao89** [comment]
recheck

**Comment by viroel** [comment]
/test dataplane-operator-build-deploy-kuttl

there were some changes in the cloud provider, lets see if we can get better results here

**Comment by fao89** [comment]
recheck

**Comment by jpodivin** [comment]
recheck 

`"Trace[2060427748]: ---\"Objects listed\" error:Unauthorized 11071ms (19:04:50.549)", "Trace[2060427748]: [11.071612908s] [11.071612908s] END", "E0123 19:04:50.549983   33198 reflector.go:148] k8s.io/client-go/tools/watch/informerwatcher.go:146: Failed to watch *unstructured.Unstructured: failed to list *unstructured.Unstructured: Unauthorized", "E0123 19:08:29.546881   33198 reflector.go:148] k8s.io/client-go/tools/watch/informerwatcher.go:146: Failed to watch *unstructured.Unstructured: unknown", "error: timed out waiting for the condition on openstackdataplanenodesets/openstack-edpm"`

**Comment by jpodivin** [comment]
recheck
`failed: [localhost] (item=cainjector) => {"ansible_loop_var": "item", "changed": false, "item": "cainjector", "module_stderr": "Traceback (most recent call last):\n  File \"/home/zuul/.local/lib/python3.9/site-packages/kubernetes/dynamic/client.py\", line 55, in inner\n    resp = func(self, *args, **kwargs)\n  File \"/home/zuul/.local/lib/python3.9/site-packages/kubernetes/dynamic/client.py\", line 270, in request\n    api_response = self.client.call_api(\n  File \"/home/zuul/.local/lib/python3.9/site-packages/kubernetes/client/api_client.py\", line 348, in call_api\n    return self.__call_api(resource_path, method,\n  File \"/home/zuul/.local/lib/python3.9/site-packages/kubernetes/client/api_client.py\", line 180, in __call_api\n    response_data = self.request(\n  File \"/home/zuul/.local/lib/python3.9/site-packages/kubernetes/client/api_client.py\", line 373, in request\n    return self.rest_client.GET(url,\n  File \"/home/zuul/.local/lib/python3.9/site-packages/kubernetes/client/rest.py\", line 244, in GET\n    return self.request(\"GET\", url,\n  File \"/home/zuul/.local/lib/python3.9/site-packages/kubernetes/client/rest.py\", line 238, in request\n    raise ApiException(http_resp=r)\nkubernetes.client.exceptions.ApiException: (403)\nReason: Forbidden\nHTTP response headers: HTTPHeaderDict({'Audit-Id': '1b135bb0-ac37-44f5-aff7-fe5818b3951f', 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains; preload', 'X-Content-Type-Options': 'nosniff', 'X-Kubernetes-Pf-Flowschema-Uid': '', 'X-Kubernetes-Pf-Prioritylevel-Uid': '', 'Date': 'Wed, 24 Jan 2024 07:25:37 GMT', 'Content-Length': '353'})\nHTTP response body: b'{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"pods \\\\\"cert-manager-cainjector-66ccc7b8fd-g5xfk\\\\\" is forbidden: User \\\\\"kubeadmin\\\\\" cannot get resource \\\\\"pods\\\\\" in API group \\\\\"\\\\\" in the namespace \\\\\"cert-manager\\\\\"\",\"reason\":\"Forbidden\",\"details\":{\"name\":\"cert-manager-cainjector-66ccc7b8fd-g5xfk\",\"kind\":\"pods\"},\"code\":403}\\n'\n\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"<stdin>\", line 107, in <module>\n  File \"<stdin>\", line 99, in _ansiballz_main\n  File \"<stdin>\", line 47, in invoke_module\n  File \"/usr/lib64/python3.9/runpy.py\", line 225, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib64/python3.9/runpy.py\", line 97, in _run_module_code\n    _run_code(code, mod_globals, init_globals,\n  File \"/usr/lib64/python3.9/runpy.py\", line 87, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_kubernetes.core.k8s_info_payload_5gwor60w/ansible_kubernetes.core.k8s_info_payload.zip/ansible_collections/kubernetes/core/plugins/modules/k8s_info.py\", line 229, in <module>\n  File \"/tmp/ansible_kubernetes.core.k8s_info_payload_5gwor60w/ansible_kubernetes.core.k8s_info_payload.zip/ansible_collections/kubernetes/core/plugins/modules/k8s_info.py\", line 223, in main\n  File \"/tmp/ansible_kubernetes.core.k8s_info_payload_5gwor60w/ansible_kubernetes.core.k8s_info_payload.zip/ansible_collections/kubernetes/core/plugins/modules/k8s_info.py\", line 183, in execute_module\n  File \"/tmp/ansible_kubernetes.core.k8s_info_payload_5gwor60w/ansible_kubernetes.core.k8s_info_payload.zip/ansible_collections/kubernetes/core/plugins/module_utils/k8s/service.py\", line 319, in find\n  File \"/tmp/ansible_kubernetes.core.k8s_info_payload_5gwor60w/ansible_kubernetes.core.k8s_info_payload.zip/ansible_collections/kubernetes/core/plugins/module_utils/k8s/waiter.py\", line 178, in wait\n  File \"/tmp/ansible_kubernetes.core.k8s_info_payload_5gwor60w/ansible_kubernetes.core.k8s_info_payload.zip/ansible_collections/kubernetes/core/plugins/module_utils/k8s/client.py\", line 318, in get\n  File \"/home/zuul/.local/lib/python3.9/site-packages/kubernetes/dynamic/client.py\", line 112, in get\n    return self.request('get', path, **kwargs)\n  File \"/home/zuul/.local/lib/python3.9/site-packages/kubernetes/dynamic/client.py\", line 57, in inner\n    raise api_exception(e)\nkubernetes.dynamic.exceptions.ForbiddenError: 403\nReason: Forbidden\nHTTP response headers: HTTPHeaderDict({'Audit-Id': '1b135bb0-ac37-44f5-aff7-fe5818b3951f', 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains; preload', 'X-Content-Type-Options': 'nosniff', 'X-Kubernetes-Pf-Flowschema-Uid': '', 'X-Kubernetes-Pf-Prioritylevel-Uid': '', 'Date': 'Wed, 24 Jan 2024 07:25:37 GMT', 'Content-Length': '353'})\nHTTP response body: b'{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"pods \\\\\"cert-manager-cainjector-66ccc7b8fd-g5xfk\\\\\" is forbidden: User \\\\\"kubeadmin\\\\\" cannot get resource \\\\\"pods\\\\\" in API group \\\\\"\\\\\" in the namespace \\\\\"cert-manager\\\\\"\",\"reason\":\"Forbidden\",\"details\":{\"name\":\"cert-manager-cainjector-66ccc7b8fd-g5xfk\",\"kind\":\"pods\"},\"code\":403}\\n'\nOriginal traceback: \n  File \"/home/zuul/.local/lib/python3.9/site-packages/kubernetes/dynamic/client.py\", line 55, in inner\n    resp = func(self, *args, **kwargs)\n\n  File \"/home/zuul/.local/lib/python3.9/site-packages/kubernetes/dynamic/client.py\", line 270, in request\n    api_response = self.client.call_api(\n\n  File \"/home/zuul/.local/lib/python3.9/site-packages/kubernetes/client/api_client.py\", line 348, in call_api\n    return self.__call_api(resource_path, method,\n\n  File \"/home/zuul/.local/lib/python3.9/site-packages/kubernetes/client/api_client.py\", line 180, in __call_api\n    response_data = self.request(\n\n  File \"/home/zuul/.local/lib/python3.9/site-packages/kubernetes/client/api_client.py\", line 373, in request\n    return self.rest_client.GET(url,\n\n  File \"/home/zuul/.local/lib/python3.9/site-packages/kubernetes/client/rest.py\", line 244, in GET\n    return self.request(\"GET\", url,\n\n  File \"/home/zuul/.local/lib/python3.9/site-packages/kubernetes/client/rest.py\", line 238, in request\n    raise ApiException(http_resp=r)\n\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}`

**Comment by jpodivin** [comment]
recheck

`retry limit`

**Pull request merged**